Replace global variables with data array in Plugin class

lukecarbis opened the pull request:
Closes #425.

lukecarbis committed:
Replace global variables with data array in Plugin class

kienstra commented:
Hi @lukecarbis,
Nice, you had a productive day!

If it's alright, tomorrow is probably the earliest I could review this.

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -332,8 +330,8 @@ public function render_block_template( $block, $attributes ) {
-		$block_lab_attributes = $attributes;
-		$block_lab_config     = $block;
+		block_lab()->data['attributes'] = $attributes;
+		block_lab()->data['config']     = $block;
```
What do you think about storing `->data` as a protected property of this `Loader` class? 

That way, it wouldn't be possible to change the data outside this class.

Then, `helpers.php` could access the data with something like `block_lab()->loader->get_data( 'attributes' )`, or whatever method name.

#440 has infrastructure for this, including storing `Loader` as a property of `Plugin`, not in the protected property `->components`.

kienstra replied:
If it's stored in the `Plugin` class, it has to be publicly accessible. But if it's stored in `Loader`, it can be set inside that class, and no other class can set it. They can only get it.

kienstra reviewed (commented):
**Question About Storing Data In Loader**

Hi @lukecarbis,
This PR looks good, and it'll keep us from being stuck with these globals.

What do you think about storing the `->data` in the `Loader` class? That should keep the data from being changed anywhere else, and it could still be accessed in `helpers.php`.

Hope you had a great weekend, it was definitely productive 😄

lukecarbis committed:
Merge branch 'develop' into feature/replace-globals

lukecarbis committed:
Move data to a protected property of Loader

lukecarbis commented:
@kienstra I've implemented your suggestion (great idea by the way), but I'm struggling with some failing integration tests. I'm kind of stuck – could you please take a look?

kienstra commented:
Sure, @lukecarbis! The integration tests aren't very easy to maintain 😄 

It might be a few days before I get to this, if that's alright.

kienstra committed:
Get the integration tests to pass, by removing extra instance of Loader

block_field() looks for the Loader in
block_lab()->loader,
but these tests were instantiating a new Loader.
So remove the instantiation.

kienstra commented on php/helpers.php:
```diff
@@ -18,14 +18,25 @@
+	 * @var array $attributes
 	 */
-	global $block_lab_attributes, $block_lab_config;
+	$attributes = block_lab()->loader->get_data( 'attributes' );
```
This DocBlock was definitely helpful when there was a global, but do you think it's needed now that it's getting `$attributes` with a normal method call?

And likewise for some other places in this PR.

lukecarbis replied:
Done in https://github.com/getblocklab/block-lab/pull/435/commits/06d78c727f1a014a6905dd05bb64d2725795ca75

kienstra commented on php/helpers.php:
```diff
@@ -126,9 +137,14 @@ function block_row( $name ) {
  * @return bool
  */
 function block_rows( $name ) {
-	global $block_lab_attributes;
```
Nice, it's great to see the `global` removed.

kienstra commented on php/helpers.php:
```diff
@@ -126,9 +137,14 @@ function block_row( $name ) {
+	$attributes = block_lab()->loader->get_data( 'attributes' );
 
-	if ( ! isset( $block_lab_attributes[ $name ] ) ) {
+	if ( ! $attributes || ! isset( $attributes[ $name ] ) ) {
```
`! isset( $attributes[ $name ] )` should be enough here, as it'll be false if `! isset( $attributes )`

lukecarbis replied:
Done in https://github.com/getblocklab/block-lab/pull/435/commits/0fbf45adb15a4ce1a8fd10112939b68a67249365

kienstra commented on php/helpers.php:
```diff
@@ -172,13 +188,18 @@ function reset_block_rows( $name ) {
+	$attributes = block_lab()->loader->get_data( 'attributes' );
 
-	if ( ! isset( $block_lab_attributes[ $name ]['rows'] ) ) {
+	if ( ! $attributes || ! isset( $attributes[ $name ]['rows'] ) ) {
```
`! isset( $attributes[ $name ]['rows'] )` should be enough, as it'll also be false if `! $attributes`

lukecarbis replied:
Done in https://github.com/getblocklab/block-lab/pull/435/commits/0fbf45adb15a4ce1a8fd10112939b68a67249365

kienstra commented on php/helpers.php:
```diff
@@ -303,9 +345,16 @@ function block_config() {
  * @return array|null
  */
 function block_field_config( $name ) {
-	global $block_lab_config;
```
It's great to see `global` removed 😄

kienstra reviewed (changes requested):
**Looks Good**
*A Few Small Points*

Hi @lukecarbis,
Nice work! This looks good.

There are just a few points.

If it's alright, I [pushed to this branch](https://github.com/getblocklab/block-lab/pull/435/commits/fe6e5835e2af38ff1bc245a6021d8aa8750cb099) to get the Integration tests to pass. Sorry, I try not to do that without asking.

Would you mind if I pushed a PHPUnit test for `Loader::get_data()` to this branch? 

Thanks, Luke!

lukecarbis commented:
@kienstra please go ahead!

kienstra committed:
Add assertions for Loader::get_data()

Test that the data is as expected,
including with filters.

kienstra commented:
Thanks, https://github.com/getblocklab/block-lab/pull/435/commits/3e143382926ff3c33da53298e570afe0726aabee adds a test for `Loader::get_data()`

lukecarbis committed:
Merge branch 'develop' into feature/replace-globals

lukecarbis committed:
Remove unneeded docblocks

lukecarbis committed:
Remove unneeded conditional statements

lukecarbis commented:
@kienstra Ready for you to take another look. :)

kienstra reviewed (approved):
**Approved**

Hi @lukecarbis,
This looks great! It's to merge when you'd like. 

It'd be good to mention in the Changelog the new API for this, like
`block_lab->loader->get_data( 'attributes' )` instead of `$block_lab_attributes`
